Add possibility to set custom properties to cart item

dblaichinger opened the pull request:
## Overview

The suggested change would make it easier to extend `CartItem` class and provide the property in `Cart::addItem()` as `$params` afterwards.

## Extend CartItem

The `$params` parameter in `AbstractCart::addItem` is currently unused in this method, but the doc-block suggests that it's possible to add additional data.

In `src/CartManager/CartInterface.php`:

>    /**
     * @param CheckoutableInterface $product
     * @param int $count
     * @param string|null $itemKey
     * @param bool $replace replace if item with same key exists
     * **@param array $params optional additional item information**
     * @param AbstractSetProductEntry[] $subProducts
     * @param string|null $comment
     *
     * @return string $itemKey
     */
    public function addItem(CheckoutableInterface $product, int $count, string $itemKey = null, bool $replace = false, array $params = [], array $subProducts = [], string $comment = null): string;

---

Also see: 
* https://github.com/pimcore/pimcore/discussions/16249
* https://github.com/pimcore/ecommerce-framework-bundle/issues/98

Commit:
Introduce setCustomProperties to CartItem which makes it possible to use $params in Cart::addItem

Commit:
Introduce setCustomProperties to CartItem which makes it possible to use $params in Cart::addItem

Commit:
Merge branch 'add-custom-properties-to-cart-item' of github.com:dblaichinger/ecommerce-framework-bundle into add-custom-properties-to-cart-item

Commit:
Merge branch 'add-custom-properties-to-cart-item' of github.com:dblaichinger/ecommerce-framework-bundle into add-custom-properties-to-cart-item

bluvulture commented:
Hey @dblaichinger 

Could you plase sign cla and then re-trigger the check?

Tnx

blankse commented on src/CartManager/AbstractCart.php:
```diff
@@ -158,6 +158,10 @@ public function updateItem(string $itemKey, CheckoutableInterface $product, int
             $item->setSubItems($subItems);
         }
 
+        if (!empty($params)) {
+            $item->setCustomProperties($params);
+        }
```
$params is always set and always an array. The if shouldn't be necessary.
```suggestion
        $item->setCustomProperties($params);
```

Commit:
Fixes review comment in AbstractCart.php

Co-authored-by: Sebastian Blank <sebastian.bl@gmx.de>

dblaichinger commented:
recheck

Commit:
Merge branch 'add-custom-properties-to-cart-item' of github.com:dblaichinger/ecommerce-framework-bundle into add-custom-properties-to-cart-item

Commit:
Merge branch 'add-custom-properties-to-cart-item' of github.com:dblaichinger/ecommerce-framework-bundle into add-custom-properties-to-cart-item

dblaichinger commented:
Hey @bluvulture,

I've agreed to the CLA, but the build is still failing...
<img width="536" alt="Screenshot 2023-12-15 at 13 41 21" src="https://github.com/pimcore/ecommerce-framework-bundle/assets/511559/7847f5de-7585-4037-bf9e-51be208ecb85">

Is there anything else I need to consider?

Thanks!

dblaichinger commented:
I have read the CLA Document and I hereby sign the CLA

Commit:
Merge branch 'add-custom-properties-to-cart-item' of github.com:dblaichinger/ecommerce-framework-bundle into add-custom-properties-to-cart-item

fashxp commented:
Can we do that without changing an interface? As it is now it would be a bc break

Commit:
Remove method declaration from CartItemInterface

dblaichinger commented:
Hi @fashxp,

I've removed the interface change in this PR and added a `method_exists` check instead.
However, I still think it's cleaner to extend the interface, so I've added a new https://github.com/pimcore/ecommerce-framework-bundle/pull/146. Can you please add the `BC-BREAK` label there and remove it from this one?

Thank you!

fashxp committed:
adding some docs

fashxp commented:
@dblaichinger thx very much. added some docs, can you please double check? 
also thought that it might be really cool, if the custom properties would be stored automatically and extending Cart/CartItem isn't even necessary. 
But that might require a bit more refactoring.

Commit:
Added more details to 11_Cart_Manager.md

dblaichinger commented:
Hi @fashxp 

> added some docs, can you please double check?

I've added some details, please check https://github.com/pimcore/ecommerce-framework-bundle/pull/143/commits/b43fa91b230f4891bfa1d4d388705e76a899caa4

> also thought that it might be really cool, if the custom properties would be stored automatically and extending Cart/CartItem isn't even necessary.

Can this be achieved without changing `CartItemInterface` and breaking compatibility?

I've added `$customProperties` to `AbstractCartItem` in the `BC-BREAK` tagged PR: https://github.com/pimcore/ecommerce-framework-bundle/pull/146/files. Maybe that's a better approach when already breaking compatibility. 

Thanks!

fashxp commented:
> I've added $customProperties to AbstractCartItem in the BC-BREAK tagged PR: https://github.com/pimcore/ecommerce-framework-bundle/pull/146/files. Maybe that's a better approach when already breaking compatibility. 

not sure if we really need to add an additional parameter, of if we just add a generic functionality to the existing cart implementations, which persists the data passed via `$params` parameter to their storage. Let's continue the discussion over there in the other PR. 

This one looks good now. Thx very much!!
